test/infrastructure/docker/exp/api/v1beta1/dockermachinepool_types.go

```diff
@@ -82,6 +85,15 @@ type DockerMachinePoolStatus struct {
 	// Conditions defines current service state of the DockerMachinePool.
 	// +optional
 	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+
+	// InfrastructureMachineSelector is a label query over the infrastructure resources behind MachinePool Machines.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	// +optional
+	InfrastructureMachineSelector metav1.LabelSelector `json:"infrastructureMachineSelector,omitempty"`
+
+	// InfrastructureMachineKind is the kind of the infrastructure resources behind MachinePool Machines.
+	// +optional
+	InfrastructureMachineKind string `json:"infrastructureMachineKind,omitempty"`
```

Comment:
Looking at this I found out that we might have a sort of inconsistency in the current API:

MachinePool.Spec.Template, is defined as "Template describes the machines that will be created", and in fact inside it we have a machine spec, bootstrap config - that defines how machine have to be bootstrapped -, but we are currently using infrastructureRef to link to the MachinePool infrastructure.
This is leading us to find a creative solutions to surface the InfrastructureMachineKind, which instead will be in infrastructureRef in all the other places where we are defining Machine templates

Reply:
I think you're right in that it's a bit odd to use `MachinePool.Spec.Template.Spec.InfrastructureRef` to point to the InfraMachinePool since I would expect it to be in a field like `MachinePool.Spec.InfrastructureRef`. But even if we had a template, it would point to InfraMachineTemplate which doesn't actually contain the name of the InfraMachineKind. 

I do agree that our current design leads to needing creative solutions to surfacing information in the various controllers. Maybe we can follow up later to find a way to unify the info in the MachinePool fields, for example, since the MachinePool resource is passed around a lot in the InfraMachinePool controllers.

Reply:
> we are currently using infrastructureRef to link to the "entire" MachinePool infrastructure

There is no such thing as "InfraMachineTemplate" for MachinePools because by definition they _are_ a template (eg. Azure VMSS has a "model", AWS ASG has a "launch template" so you could argue that the InfraMachinePool itself is the best equivalent to a template (and in itself the InfraMachinePool isn't a piece of infra, but a logical grouping of scalable machines). So just I think its fair to say that `InfraMachinePool` is to `MachinePool` what `InfraMachineTemplate` is to `MachineDeployment` (note that there is no such thing as a `InfraMachineDeployment`).

If you look at other equivalents in the code, `MachineDeployments` has an infraRef to the InfraMachineTemplate, not the InfraMachine itself, so there is no way to get the InfraMachine. kind from the `MachineDeployment` either. The only place where you'll find the InfraMachine as an InfrastructureRef is in the `Machine` itself, which we can't use here because they don't exist yet.

Reply:
> breaks the promises of using template as a description of the machines the will be created

I agree with this part. You could argue that the API is malformed since the `MachinePool` controller isn't responsible for creating "clones" of `AzureMachinePools` (the cloud provider does that), so maybe it shouldn't be called "template". We could consider changing the `MachinePool` spec to take the infraRef and bootstrapRef out of the "template" field (and get rid of the template field altogether). However, I would say this is out of scope of this specific PR as the API for MachinePool pre-exists this change and changing it won't have an impact on this PR. probably something we need to figure out before experimental graduation though since it is a breaking change.